java/src/jmri/Block.java

```diff
@@ -1074,15 +1095,16 @@ public void goingActive() {
 
                 if (next != null && count == 1) {
                     // found one block with proper direction, use it
-                    setValue(next.getBlock().getValue());
+                    Object nextBlockValue = getBlockValueFromPath(next);
+                    setValue(nextBlockValue);
                     setDirection(next.getFromBlockDirection());
                     log.debug("Block {} gets new value '{}' from {}, direction {}",
-                            getDisplayName(), next.getBlock().getValue(),
+                            getDisplayName(), nextBlockValue,
                             next.getBlock().getDisplayName(), Path.decodeDirection(getDirection()));
                 } else {
                     // handle merging trains: All neighbors with same content (train ID)
                     if (allNeighborsAgree && next != null) {
-                        setValue(next.getBlock().getValue());
+                        setValue(getBlockValueFromPath(next));
                         setDirection(next.getFromBlockDirection());
```

Comment:
Before this change, `next.getBlock().getValue()` would cause a NPE if `next.getBlock()` is null. With this change, `setValue()` will be called with `null`. Maybe the `if` statement should be:

`if (allNeighborsAgree && next != null && next.getBlock() != null) {`

Reply:
This class is messy and doing complicated things; I'm not sure I understand it all.  But it seems to me that `allNeighborsAgree` can't be true if the next block has a null value. `allNeighborsAgree` means that the neighboring blocks all have the same _non_null_ value, which allows the code to assume that the current block should have that value to.  E.g. detection was temporarily lost for a block in the middle of the train.